R/pkg/R/serialize.R

```diff
@@ -37,24 +37,38 @@ writeObject <- function(con, object, writeType = TRUE) {
   # passing in vectors as arrays and instead require arrays to be passed
   # as lists.
   type <- class(object)[[1]]  # class of POSIXlt is c("POSIXlt", "POSIXt")
-  if (writeType) {
-    writeType(con, type)
+  isNA <- FALSE
+  if (type %in% c("integer", "character", "logical", "double", "numeric")) {
+    isNA <- writeNA(con, object)
   }
-  switch(type,
-         NULL = writeVoid(con),
-         integer = writeInt(con, object),
-         character = writeString(con, object),
-         logical = writeBoolean(con, object),
-         double = writeDouble(con, object),
-         numeric = writeDouble(con, object),
-         raw = writeRaw(con, object),
-         list = writeList(con, object),
-         jobj = writeJobj(con, object),
-         environment = writeEnv(con, object),
-         Date = writeDate(con, object),
-         POSIXlt = writeTime(con, object),
-         POSIXct = writeTime(con, object),
-         stop(paste("Unsupported type for serialization", type)))
+  if (!isNA) {
+    if (writeType) {
+      writeType(con, type)
+    }
+    switch(type,
+           NULL = writeVoid(con),
+           integer = writeInt(con, object),
+           character = writeString(con, object),
+           logical = writeBoolean(con, object),
+           double = writeDouble(con, object),
+           numeric = writeDouble(con, object),
+           raw = writeRaw(con, object),
+           list = writeList(con, object),
+           jobj = writeJobj(con, object),
+           environment = writeEnv(con, object),
+           Date = writeDate(con, object),
+           POSIXlt = writeTime(con, object),
+           POSIXct = writeTime(con, object),
+           stop(paste("Unsupported type for serialization", type)))
+  }
+}
+
+writeNA <- function(con, value) {
+  if(is.na(value)) {
+    writeBin(charToRaw("n"), con) 
```

Comment:
Could we instead use `writeType(NULL)` ? If that is equivalent it will be more readable